### ISSUE TITLE
I want to add a metric type for `sum distinct`

### ISSUE THREAD

**Issue by TuringLovesDeathMetal** [lightdash/lightdash#4329]
### Description

I want to be able to create a sum of values that are based on a distinct list of items. 

For example, I have a table like this:
<img width="409" alt="image" src="https://user-images.githubusercontent.com/31848148/215535028-3b94434e-e513-41a6-a154-01785fe6c1dd.png">

I want to create a sum of all item costs, based on the distinct item number. So, my `sum_distinct` for this table would be 15. 

I would have to define the field I wanted to be `distinct` on for the metric, then define the dimension I'm summing. So the .yml could look something like this:

```yml
name: cost
  meta:
    metrics:
      sum_distinct_cost:
        type: sum_distinct
        distinct_keys: 
           - ${item_number}
           - ${item_name}
```

The SQL for this metric would look something like this (we'd probably need to use hashing functions):

```sql

SELECT 
SUM(DISTINCT(cost + HASH(CONCAT(${item_number), ${item_name})))) - SUM(DISTINCT(HASH(CONCAT(${item_number), ${item_name}))) AS sum_distinct_cost

FROM table
```

### Acceptance criteria
- [ ] I'm able to add `sum distinct` metrics in my .yml
- [ ] I can define a sum distinct and specify which columns it needs to be distinct on
- [ ] I can reference columns from joined tables in my `distinct_key` values
- [ ] the `distinct_key` can take a list or a single value
- [ ] docs have been updated to include the sum distinct metric: https://docs.lightdash.com/references/metrics#metric-types

**Comment by 12ian34** [comment]
Just requested by a partner.

**Comment by DougsHub** [comment]
I'd love to request this feature as well!

**Comment by fev4** [comment]
this would be such a great addition!

**Comment by GaylorBrunner22** [comment]
Very relevant, I want it too !

**Comment by yu-iskw** [comment]
@TuringLovesDeathMetal told me a workaround to calculate something like `sum_distinct` using the symmetric aggregates approach. However, I found it difficult to precisely calculate it.

## TL;DR

I figured out that the symmetric aggregation on Lightdash with BigQuery is difficult, because overflow happens when summing the values.

## Background

- As far as I know, the `FARM_FINGERPRINT` function is the only built-in function to convert a value to a hash of number.

## What I tried

I tried the symmetric aggregation with the `orders` table of the `jaffle_shop` dataset.
Here, I implement the metrics to virtually calculate the sum of the `amount` column on unique `order_id` values.
If we just sum the values of the `amount` column on unique `order_id` values, the value is `1672.0`.

### Step 1: Basic Implementation

I started with the following basic implementation.
However, I encountered the integer overflow error: `Error in SUM aggregation: integer overflow` on the Lightdash UI.

```yaml
meta:
  metrics:
    sum_distinct_order_amount:
      type: number
      sql: |
        ${sum_distinct_order_amount_helper1} - ${sum_distinct_order_amount_helper2}
    sum_distinct_order_amount_helper1:
      type: sum
      hidden: true
      sql: |
        DISTINCT(${amount} + FARM_FINGERPRINT(${order_id}))
    sum_distinct_order_amount_helper2:
      type: sum
      hidden: true
      sql: |
        DISTINCT(FARM_FINGERPRINT(${order_id}))
```

## Step 2: Casting to NUMERIC

So, I tried casting the values to NUMERIC to avoid the overflow issue.
However, I got a resulting negative value as `-12288`.

According to my research, the value of the `amount` column is relatively too small against the values of the `FARM_FINGERPRINT` function.
So, the calculation result can be incorrect.

```yaml
meta:
  metrics:
    sum_distinct_order_amount_helper1:
      type: sum
      hidden: true
      sql: |
        DISTINCT(${amount} + SAFE_CAST(FARM_FINGERPRINT(${order_id}) AS NUMERIC))
    sum_distinct_order_amount_helper2:
      type: sum
      hidden: true
      sql: |
        DISTINCT(SAFE_CAST(FARM_FINGERPRINT(${order_id}) AS NUMERIC))
```

### Step 3: Using smaller hash values

I assumed that the value of the `FARM_FINGERPRINT` function is too large to the values of the `amount` column.
So, I divided the resulting values to make the value smaller.
Finally, I got `1671.9999389648438` similar to the actual value of the summation.

However, if the results by dividing the values are not sure to unique `order_id` values.
If the uniqueness is not guaranteed, the results are not sure to be correct.
Meanwhile, the value to divide the resulting values is depends on the value of the `amount` column.
We probably need to find the appropriate value to divide the resulting values.

```yaml
meta:
  metrics:
    sum_distinct_order_amount_helper1:
      type: sum
      hidden: true
      sql: |
        DISTINCT(${amount} + SAFE_CAST(FARM_FINGERPRINT(${order_id}) AS NUMERIC) / 123456789.0)
    sum_distinct_order_amount_helper2:
      type: sum
      hidden: true
      sql: |
        DISTINCT(SAFE_CAST(FARM_FINGERPRINT(${order_id}) AS NUMERIC) / 123456789.0)
```

**Comment by yu-iskw** [comment]
## A workaround to virtually calculate sum distinct count in Lightdash

I figured out a workaround to virtually implement sum distinct count in Lightdash with BigQuery.
We assume if we have the subsequent data with ARRAY type column.
We have to expand the ARRAY field into separate rows so that we can deal with each element as a row in Lightdash, because Lightdash doesn't natively support ARRAY type columns.
So, It causes the issue to calculate summation metrics on duplicated rows.

BigQuery offers [`ROW_NUMBER()` window function](<https://cloud.google.com/bigquery/docs/reference/standard-sql/numbering_functions#row_number>), which can be used as a unique ID within each partition.
We can select only the first row of each expanded rows to avoid the issue.

1. We add a Lightdash dimension with `ROW_NUMBER()` window function to the ARRAY column.
2. We set a filter on the Lightdash UI to select only the first row of each expanded rows.

There are two approaches to add a dimension with `ROW_NUMBER()` window function.
First, we can implement a dimension with `ROW_NUMBER()` in the target dbt model.
Second, we can add a dimension with `ROW_NUMBER()` as an additional dimension in the Lightdash UI.

![image](https://github.com/user-attachments/assets/12a57d13-9097-46e8-beaf-640e5cdf0cda)


```sql
-- The original data doesn't have duplicated rows
WITH jobs AS (
  SELECT
    1 AS job_id,
    TIMESTAMP('2024-01-01T00:00:00Z') AS creation_time,
    ARRAY<STRING>['a', 'b'] AS tags,
  UNION ALL
  SELECT
    2 AS job_id,
    TIMESTAMP('2024-01-02T00:00:00Z') AS creation_time,
    ARRAY<STRING>['a', 'c'] AS tags,
  UNION ALL
  SELECT
    3 AS job_id,
    TIMESTAMP('2024-01-03T00:00:00Z') AS creation_time,
    ARRAY<STRING>[] AS tags,
)

-- We have to expand the ARRAY field into separate rows by the tag.
-- That means we have to create duplicate rows for the same job_id.
SELECT
  ROW_NUMBER() OVER (PARTITION BY jobs.job_id ORDER BY jobs.creation_time) AS job_id_row_number,
  jobs.job_id,
  jobs.creation_time,
  tag,
FROM jobs AS jobs
LEFT JOIN UNNEST(jobs.tags) AS tag
```

```yaml
    columns:
      - name: job_id_row_number
        meta:
          dimension:
            type: number
            label: job_id_row_number
```
